Add HasCallStack to functions (resolves #40)

tbidne opened the pull request:
Resolves #40. I'm putting this up now for visibility, though it's probably best to wait until GHC 9.6 can be added to CI so the new code can actually be compiled on CI (9.6 is the [first GHC](https://gitlab.haskell.org/ghc/ghc/-/wikis/commentary/libraries/version-history) using `exceptions >= 0.10.6` that this PR utilizes).

Couple notes:

1. In addition to the `MIN_VERSION_base(4,9,0)` guard for `HasCallStack`, I added a condition for `MIN_VERSION_exceptions(0,10,6)`, since there is no reason to add `HasCallStack` for any lower versions (also it would trip `-Wredundant-constraints`, though that doesn't appear to be on).

2. It is not entirely clear to me when it is best to use `withFrozenCallStack`, so I followed the [original PR's](https://github.com/ekmett/exceptions/pull/90/files) lead. The following table summarizes the situation:

    * LHS listing `exceptions` functions using `withFrozenCallStack`
    * RHS is either the `safe-exceptions` counterpart, or a note that the original `exceptions` function was re-exported, so we do not have to do anything.
    
    Additionally, `safe-exceptions` has a few functions that do not exist in `exceptions`, yet seem they like they should also use `withFrozenCallStack`, so I added those as well, and listed them as `N/A`.

    <details>
        <summary>Click to expand table</summary>

    | `exceptions`           | `safe-exceptions`  |
    |------------------------|--------------------|
    | `mask_`                | Re-exported        |
    | `uninterruptibleMask_` | Re-exported        |
    | `catchAll`             | `catchAny`         |
    | `catchIOError`         | Re-exported        |
    | `catchIf`              | N/A                |
    | `catchJust`            | `catchJust`        |
    | `handle`               | `handle`           |
    | `handleIOError`        | Re-exported        |
    | `handleAll`            | `handleAny`        |
    | `handleIf`             | N/A                |
    | `handleJust`           | `handleJust`       |
    | `try`                  | `try`              |
    | `tryJust`              | `tryJust`          |
    | `catches`              | `catches`          |
    | `onException`          | `onException`      |
    | `onError`              | N/A                |
    | `bracket`              | `bracket`          |
    | `bracket_`             | `bracket_`         |
    | `finally`              | `finally`          |
    | `bracketOnError`       | Re-exported        |
    | N/A                    | `catchIO`          |
    | N/A                    | `catchDeep`        |
    | N/A                    | `catchAnyDeep`     |
    | N/A                    | `handleIO`         |
    | N/A                    | `handleAny`        |
    | N/A                    | `handleDeep`       |
    | N/A                    | `handleAnyDeep`    |
    | N/A                    | `tryIO`            |
    | N/A                    | `tryAny`           |
    | N/A                    | `tryDeep`          |
    | N/A                    | `tryAnyDeep`       |
    | N/A                    | `withException`    |
    | N/A                    | `bracketOnError_`  |
    | N/A                    | `bracketWithError` |
    | N/A                    | `catchesDeep`      |

    </details>

Perhaps @parsonsmatt would be kind enough to glance over the `withFrozenCallStack` logic?

tbidne committed:
Add HasCallStack to functions (resolves #40)

parsonsmatt commented on src/Control/Exception/Safe.hs:
```diff
@@ -102,22 +102,65 @@ import GHC.Stack (prettySrcLoc)
 --
 -- @since 0.1.0.0
-throw :: (C.MonadThrow m, Exception e) => e -> m a
+throw :: HAS_CALL_STACK => (C.MonadThrow m, Exception e) => e -> m a
 throw = C.throwM . toSyncException
```
We *could* put a `withFrozenCallStack` here, which would "hide" that `Control.Monad.Catch.throwM` is the underlying throwing mechanism. Not sure if that's desired or not.

parsonsmatt commented on src/Control/Exception/Safe.hs:
```diff
@@ -102,22 +102,65 @@ import GHC.Stack (prettySrcLoc)
 --
 -- @since 0.1.0.0
-throwIO :: (C.MonadThrow m, Exception e) => e -> m a
+throwIO :: HAS_CALL_STACK => (C.MonadThrow m, Exception e) => e -> m a
 throwIO = throw
```
I'd definitely recommend `withFrozenCallStack` here, these are just aliases

```suggestion
throwIO :: HAS_CALL_STACK => (C.MonadThrow m, Exception e) => e -> m a
throwIO = withFrozenCallStack throw
```

parsonsmatt commented on src/Control/Exception/Safe.hs:
```diff
@@ -102,22 +102,65 @@ import GHC.Stack (prettySrcLoc)
 --
 -- @since 0.1.0.0
-throwM :: (C.MonadThrow m, Exception e) => e -> m a
+throwM :: HAS_CALL_STACK => (C.MonadThrow m, Exception e) => e -> m a
 throwM = throw
```
```suggestion
throwM :: HAS_CALL_STACK => (C.MonadThrow m, Exception e) => e -> m a
throwM = withFrozenCallStack throw
```

parsonsmatt commented on src/Control/Exception/Safe.hs:
```diff
@@ -183,7 +226,7 @@ instance Exception StringException
 --
 -- @since 0.1.0.0
-throwTo :: (Exception e, MonadIO m) => ThreadId -> e -> m ()
+throwTo :: HAS_CALL_STACK => (Exception e, MonadIO m) => ThreadId -> e -> m ()
 throwTo tid = liftIO . E.throwTo tid . toAsyncException
```
This one is probably unnecssary, since `E.throwTo` doesn't do anything with `CallStack` information. It also may not, depending on how the GHC proposal goes.

parsonsmatt commented on src/Control/Exception/Safe.hs:
```diff
@@ -247,47 +290,47 @@ catchAnyDeep = catchDeep
-catchJust :: (C.MonadCatch m, Exception e) => (e -> Maybe b) -> m a -> (b -> m a) -> m a
-catchJust f a b = a `catch` \e -> maybe (throwM e) b $ f e
+catchJust :: HAS_CALL_STACK => (C.MonadCatch m, Exception e) => (e -> Maybe b) -> m a -> (b -> m a) -> m a
+catchJust f a b = withFrozenCallStack a `catch` \e -> maybe (throwM e) b $ f e
```
This freezes the `CallStack` for `a`, which is probably not what you want - I'd guess you want to freeze the callstack for `catch`, which would require:

```suggestion
catchJust f a b = withFrozenCallStack catch a (\e -> maybe (throwM e) b $ f e)
```

parsonsmatt commented on src/Control/Exception/Safe.hs:
```diff
@@ -344,28 +387,28 @@ tryAnyDeep = tryDeep
 -- @since 0.1.0.0
-withException :: (C.MonadMask m, E.Exception e) => m a -> (e -> m b) -> m a
-withException thing after = C.uninterruptibleMask $ \restore -> do
+withException :: HAS_CALL_STACK => (C.MonadMask m, E.Exception e) => m a -> (e -> m b) -> m a
+withException thing after = withFrozenCallStack $ C.uninterruptibleMask $ \restore -> do
```
There's a subtle thing here - `withFrozenCallStack $` is going to freeze the `CallStack` for `thing` and `after`, while `withFrozenCallStack C.uninterruptibleMask` is going to only freeze the `CallStack` for that function. I suspect we want the former, though I *also* suspect that it's going to play weirdly with the types. May be worth not freezing here.

tbidne replied:
>  I suspect we want the former

Did you mean to say `latter` here, by any chance? If so I agree with your comment: `withFrozenCallStack C.uninterruptibleMask` seems closer to what we want (freeze the handler only), and it is consistent with other functions. And you are right about it misbehaving. Naively changing it to

```haskell
withFrozenCallStack C.uninterruptibleMask $ \restore
```
gives

```
• Couldn't match type: m a -> m a
                 with: forall a1. m a1 -> m a1
  Expected: (forall a1. m a1 -> m a1) -> m a
    Actual: (m a -> m a) -> m a
```

tbidne replied:
Fun fact, ghc can be placated with `-XImpredicativeTypes`

```haskell
-- requires -XImpredicativeTypes
uninterruptibleMaskFrozen :: forall m b. C.MonadMask m => ((forall a. m a -> m a) -> m b) -> m b
uninterruptibleMaskFrozen = withFrozenCallStack C.uninterruptibleMask

withException thing after = uninterruptibleMaskFrozen $ \restore -> do ...
```

I'm not sure how to do this without it, however (`-XRankNTypes` is not enough).

parsonsmatt commented on src/Control/Exception/Safe.hs:
```diff
@@ -534,15 +577,15 @@ displayException = show
 -- @since 0.1.2.0
-catches :: (C.MonadCatch m, C.MonadThrow m) => m a -> [Handler m a] -> m a
-catches io handlers = io `catch` catchesHandler handlers
+catches :: HAS_CALL_STACK => (C.MonadCatch m, C.MonadThrow m) => m a -> [Handler m a] -> m a
+catches io handlers = withFrozenCallStack io `catch` catchesHandler handlers
```
```suggestion
catches :: HAS_CALL_STACK => (C.MonadCatch m, C.MonadThrow m) => m a -> [Handler m a] -> m a
catches io handlers = withFrozenCallStack catch io (catchesHandler handlers)
```

parsonsmatt commented on src/Control/Exception/Safe.hs:
```diff
@@ -534,15 +577,15 @@ displayException = show
 -- @since 0.1.2.0
-catchesDeep :: (C.MonadCatch m, C.MonadThrow m, MonadIO m, NFData a) => m a -> [Handler m a] -> m a
-catchesDeep io handlers = evaluateDeep io `catch` catchesHandler handlers
+catchesDeep :: HAS_CALL_STACK => (C.MonadCatch m, C.MonadThrow m, MonadIO m, NFData a) => m a -> [Handler m a] -> m a
+catchesDeep io handlers = withFrozenCallStack evaluateDeep io `catch` catchesHandler handlers
```
```suggestion
catchesDeep :: HAS_CALL_STACK => (C.MonadCatch m, C.MonadThrow m, MonadIO m, NFData a) => m a -> [Handler m a] -> m a
catchesDeep io handlers = withFrozenCallStack catch (evaluateDeep io) (catchesHandler handlers)
```

parsonsmatt reviewed (commented):
There are a few places that I'd change the frozen call stack behavior. It's a tricky function to apply correctly - IMO it's almost easier to write with parens explicitly to make it clearer what's happening.

```haskell
withFrozenCallStack foo bar baz
withFrozenCallStack $ foo bar baz
(withFrozenCallStack foo) bar baz
withFrozenCallStack (foo bar baz)
```

Usually these are equivalent, but the implicit param obscures how these differ.

tbidne committed:
PR feedback

- Add withFrozenCallStack to simple aliases.
- Remove HasCallStack from throwTo as it is unused.
- Rewrite functions such that withFrozenCallStack is applied to the
  handler only (e.g. catch), not the action.
- Remove withFrozenCallStack from withException as desired
  semantics are unclear.

tbidne commented:
@parsonsmatt Thanks for the review! Indeed, "`withFrozenCallStack` best practices" are not totally clear to me, so your comments are super helpful. I implemented most of your requests. Some comments:

1. I am indifferent on `throw`.
1. I removed `withFrozenCallStack` from `withException` due to the uncertainty + type problems. Better to leave some noise in than to accidentally throw useful information away. Of course I am open to adding it back if there is a good argument in its favor.

The same problem for `withException` (described [here](https://github.com/fpco/safe-exceptions/pull/41#discussion_r1185568012)) exists for `finally`, so whatever is decided for `withException`, the same should apply to `finally`.

tbidne committed:
Remove withFrozenCallStack from finally

Removes withFrozenCallStack from finally for consistency with withException.
Adds explanatory comment.

tbidne committed:
Add recent snapshots to CI

Also remove windows CI workaround now that issue has been resolved.

tbidne commented:
Hey @snoyberg, I believe this is ready for your review, now that stackage nightly has 9.6 and thus this PR can be tested on CI.

In addition to the `HasCallStack` changes, I took the liberty of:

* Adding recent snapshots to CI and removing your windows workaround as the referenced issue is marked resolved (this [passed](https://github.com/tbidne/safe-exceptions/actions/runs/5352738775) on my fork, though CI needs to be authorized to run here).
* Bumped the patch.

Let me know if you want something else. Thanks!

snoyberg reviewed (approved):
LGTM. @parsonsmatt mind reviewing before I merge/release?

parsonsmatt commented on src/Control/Exception/Safe.hs:
```diff
@@ -344,27 +387,56 @@ tryAnyDeep = tryDeep
+onException :: HAS_CALL_STACK => C.MonadMask m => m a -> m b -> m a
+onException thing after = withFrozenCallStack withException thing (\(_ :: SomeException) -> after)
+
+-- Note: [withFrozenCallStack impredicativity]
```
I'm curious if `DeepSubsumption` helps here? https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0511-deep-subsumption.rst

tbidne replied:
@parsonsmatt Good idea! Unfortunately, I just tried it and received the same error:

```haskell
{-# LANGUAGE DeepSubsumption #-}
...
withException thing after = withFrozenCallStack C.uninterruptibleMask $ ...

-- src/Control/Exception/Safe.hs:441:49: error: [GHC-83865]
--     • Couldn't match type: m a -> m a
--                      with: forall a1. m a1 -> m a1
--       Expected: (forall a1. m a1 -> m a1) -> m a
--         Actual: (m a -> m a) -> m a
--     • In the first argument of ‘withFrozenCallStack’, namely
--         ‘C.uninterruptibleMask’
```

This however inspired me to try manually eta-expanding All The Things, and it actually worked without any new extensions!

```diff
diff --git a/src/Control/Exception/Safe.hs b/src/Control/Exception/Safe.hs
index 0c4b5e9..8872d3c 100644
--- a/src/Control/Exception/Safe.hs
+++ b/src/Control/Exception/Safe.hs
@@ -437,7 +437,7 @@ onException thing after = withFrozenCallStack withException thing (\(_ :: SomeEx
 --
 -- @since 0.1.0.0
 withException :: HAS_CALL_STACK => (C.MonadMask m, E.Exception e) => m a -> (e -> m b) -> m a
-withException thing after = C.uninterruptibleMask $ \restore -> do
+withException thing after = withFrozenCallStack (\io -> C.uninterruptibleMask (\r -> io r)) $ \restore -> do
     fmap fst $ C.generalBracket (pure ()) cAfter (const $ restore thing)
   where
     -- ignore the exception from after, see bracket for explanation
@@ -464,7 +464,7 @@ bracket_ before after thing = withFrozenCallStack bracket before (const after) (
 --
 -- @since 0.1.0.0
 finally :: HAS_CALL_STACK => C.MonadMask m => m a -> m b -> m a
-finally thing after = C.uninterruptibleMask $ \restore -> do
+finally thing after = withFrozenCallStack (\io -> C.uninterruptibleMask (\r -> io r)) $ \restore -> do
     fmap fst $ C.generalBracket (pure ()) cAfter (const $ restore thing)
   where
     -- ignore the exception from after, see bracket for explanation
```

I missed this when I was experimenting earlier. I am curious what others think of this idea.

parsonsmatt reviewed (approved):
Looks good to me! One curiosity re deep subumption on the withFrozenCallStack note, but I'd be happy to just merge/release as is

tbidne commented:
Thanks everyone!
